### ISSUE TITLE
[off-topic]: public vs internal package naming

### ISSUE THREAD

**Issue by glerchundi** [uber/prototool#469]
This is not a neither feature request nor issue of this fantastic tool, congrats by the way.

But it's more a question of how to handle public vs internal package organization / naming.

You state in [this documentation block](https://github.com/uber/prototool/tree/dev/style#package-naming) that packages should never include any of those reserved names (including `internal`) but recently we encountered a use case in our company that the 'same' entity (same storage), is rendered differently depending on where is this coming from (Public API or Another Internal Microservice).

So the question I would like to be answered is how do you handle this problem at Uber? Do you have this problem at all?

Why not:
* `uber.ride.v1`
* `uber.ride.vinternal`?
* `uber.ride-internal`?

Please don't hesitate to close this issue as it's clearly an off-topic but I didn't know which was the best forum to ask this one.

Thanks in advance.

**Comment by smaye81** [comment]
We have had a problem very similar.  One way we've done this in the past is something like:

`uber.foo.v1`
`uber.fooprivate.v1`

It's not great, but it's how we've handled it without having to deal with reserved words.

**Comment by glerchundi** [comment]
Thanks for your time @smaye81!

We finally reached to a solution we feel comfortable with. It's based on three ideas:
* Create a new front facing container with only the publicly available fields and methods
* Don't change the namespace so that non-existing fields are considered not set for the internal service thus making it wire-compatible.
* Keep public and internal spec in sync by automating the process (allowing us to maintain only the internal spec). This required us to create a `.proto` to `.proto` shrinker based on a custom fields&method extension: `mycompany.internal`.

I took the liberty of doing a diagram:

![Sketch - 6](https://user-images.githubusercontent.com/2232214/60824093-4921af00-a1a9-11e9-9b1a-a7939f7edeb0.png)

It adds an extra hop but the benefits outweights the cons.
